# This is SECTION 1 of the assignment

# Importing BeautifulSoup and URL Lib libraries to read the wikipedia page and scrap the table

In [2]:
import urllib.request
from bs4 import BeautifulSoup

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
page = urllib.request.urlopen(url)

# Now the content of the wikipedia page is stored in 'page' variable. Now we will use BeautifulSoup function to read the html data in soup varaible below

In [4]:
soup = BeautifulSoup(page)

# Now we will move all the tables in the page with html tag of table to below variable

In [5]:
all_tables=soup.find_all("table")

# From the 'all_table' we will find only the required table of postalcode, borough and neighborhood into another variable, filtering it further.

In [6]:
right_table=soup.find('table', class_='wikitable sortable')

# Now we will move the values from html tags to three different arrays as we have three columns

In [7]:
A=[]
B=[]
C=[]

for row in right_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==3:
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))

# Moving this to panda dataframe

In [8]:
import pandas as pd
df=pd.DataFrame(A,columns=['PostalCode'])
df['Borough']=B
df['Neighborhood']=C

df

PostalCode             Borough  \
0        M1A\n      Not assigned\n   
1        M2A\n      Not assigned\n   
2        M3A\n        North York\n   
3        M4A\n        North York\n   
4        M5A\n  Downtown Toronto\n   
..         ...                 ...   
175      M5Z\n      Not assigned\n   
176      M6Z\n      Not assigned\n   
177      M7Z\n      Not assigned\n   
178      M8Z\n         Etobicoke\n   
179      M9Z\n      Not assigned\n   

                                          Neighborhood  
0                                                   \n  
1                                                   \n  
2                                          Parkwoods\n  
3                                   Victoria Village\n  
4                          Regent Park, Harbourfront\n  
..                                                 ...  
175                                                 \n  
176                                                 \n  
177                                                 \n  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                                 \n  

[180 rows x 3 columns]

# Cleaning up the values in each row and removing Boroughs with Not assigned values as instructed.

In [9]:
df['PostalCode'] = df['PostalCode'].str.replace('\n','')
df['Borough'] = df['Borough'].str.replace('\n','')
df['Neighborhood'] = df['Neighborhood'].str.replace('\n','')
df = df[df.Borough != "Not assigned"]
df = df.reset_index(drop=True)
df.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

# Using shape method to get the total number of rows

In [11]:
df.shape

(103, 3)

# There are 103 rows in the dataframe after clean-up.

# This is SECTION 2 of the assignment